src/storage/src/source/reclock.rs

```diff
@@ -8,413 +8,795 @@
-
-            while let Some(v) = snap_iter.next().await {
-                for ((key, value), timestamp, diff) in v {
-                    let _: () = key.unwrap();
-                    let partition = value.unwrap();
-                    let (current_ts, current_offset) = snapshot_map
-                        .entry(partition)
-                        .or_insert((0, MzOffset::default()));
-
-                    *current_offset += diff;
-                    *current_ts = timestamp;
+        let mut operator = Self {
+            remap_trace: HashMap::new(),
+            since: as_of.clone(),
+            upper: Antichain::from_elem(Timestamp::minimum()),
+            source_upper: HashMap::new(),
+            write_handle,
+            read_handle,
+            listener,
+            now,
+            update_interval_ms: u64::try_from(update_interval.as_millis()).expect("huge duration"),
+        };
+
+        // Load the initial state that might exist in the shard
+        operator.sync(&upper).await;
+
+        Ok(operator)
+    }
+    /// Reclocks a batch of messages timestamped with `SourceTime` and returns an iterator of
+    /// messages timestamped with `DestTime`.
+    ///
+    /// The returned iterator will drain the provided batch as it being consumed. It is not
+    /// guaranteed that the provided batch will be cleared if the iterator is dropped in a
+    /// partially consumed state.
+    ///
+    /// The method returns an error if any of the messages is timestamped at a `SourceTime` that is
+    /// not beyond the since frontier. The error will contain the offending `SourceTime`.
+    pub async fn reclock<'a, M>(
+        &'a mut self,
+        batch: &'a mut HashMap<PartitionId, Vec<(M, MzOffset)>>,
+    ) -> Result<ReclockIter<'a, M>, (PartitionId, MzOffset)> {
+        let mut batch_upper = HashMap::with_capacity(batch.len());
+        for (pid, messages) in batch.iter_mut() {
+            messages.sort_unstable_by(|a, b| a.1.cmp(&b.1));
+            if let Some((_msg, offset)) = messages.first() {
+                let part_since = self.partition_since(pid);
+                if !(part_since <= *offset) {
+                    return Err((pid.clone(), *offset));
                 }
             }
+            if let Some((_msg, offset)) = messages.last() {
+                batch_upper.insert(pid, *offset + 1);
+            }
+        }
 
-            snapshot_map
-                .into_iter()
-                .map(|(partition, (ts, offset))| {
-                    (
-                        partition,
-                        VecDeque::from_iter(std::iter::once((ts, offset))),
-                    )
-                })
-                .collect()
-        } else {
-            HashMap::new()
-        };
-        let read_cursors = HashMap::new();
+        // Ensure we have enough bindings and re-sync state if we lost a race
+        while let Err(Upper(global_upper)) = self.mint(&batch_upper).await {
```

Comment:
It sounds like in your example `R1` must always be making progress for `R0` to seemingly get stuck, but the system as a whole is making progress because the offsets did in fact get reclocked and not only that the frontier is constantly being pushed forward by `R1` (otherwise `R0` wouldn't be losing the races).

This means that an external observer of STORAGE (e.g COMPUTE) will be seeing the ingestion progressing normally. Does this match your understanding?

I think the real question is what should a `storaged` process do when it discovers a race. When a `compare_and_append` fails at the end of the pipeline where we write to the final persist collection we currently panic and crash. This enough to ensure that we don't end up with a bunch of `storaged` instances all competing with each other so it's unclear if we should do the same in all `compare_and_append` operation. One reason to *not* do it is that it is possible to race with yourself. If you to a `compare_and_append` that succeeds on the persist side but due to networks errors comes back as an `ExternalError` then we'll retry and fine that the upper has moved and initiate a sync. In the "raced with myself" scenario this is the correct thing to do and not to crash. Crashing however is also fine, the orchestrator will notice that we're out of storaged processes and schedule a new one shortly. Shrug.

Reply:
> this extra bit of code wouldn't add much to ensuring global progress

Right -- to be clear, my argument doesn't concern global progress.  My argument is that the more sure we are that an individual replica will always make progress, the easier it is to reason about the correctness of the system.   It's not an argument about correctness (I think the code is correct!) but one about code readability.

"we have N replicas that, no matter what, will make progress so we know we must in fact make global progress" is more straightforward than "individual replicas may or may not make progress and could in fact get stuck in (albeit rare) loops but we have convinced ourselves that those situations won't turn into an issue of not making global progress".  You're losing an amount of modularity -- which I think is the biggest factor in people whether people can understand and reason about the system enough to confidently make changes

Reply:
I still disagree that this would lead to easier to understand logic, for me if statements are always better if they don't exist. 

In order to resolve this conversation I created new `advance` function that loops forever until the upper is advanced as dictated by `next_mint_timestamp` and made `reclock` only add to the remap collection when bindings need to be minted. These are now called at the appropriate times by the source ingestion code. I put the diff in a separate commit https://github.com/MaterializeInc/materialize/pull/13123/commits/1688ba78f8b0d662fb64c1b7e1ce11860e47a0cb